man/man1/ccdslice.1
.TH CCDSLICE 1NEMO "26 August 2024"

.SH "NAME"
ccdslice \- takes slices and/or re-orient an image cube

.SH "SYNOPSIS"
\fBccdslice\fP [parameter=value] ...

.SH "DESCRIPTION"
\fBccdslice\fP takes slices from an image cube, along any (orthogonal)
axis and can thereby also reorder the axes, or remove a redundant axis.
.PP
A chain of \fBccdslice\fP has to be applied if
the first (x) or second (y) axis needs to be reduced but the ordering
of the cube coordinates to remain the same (see EXAMPLES below),
see also \fIccdsub(1NEMO)\fP for a better alternative.
.PP
\fIccdfits(1NEMO)\fP can also remove redundant ("dummy") axes
when writing out a \fIfits(5NEMO)\fP file.
.PP
Using the \fBzslabs=\fP keyword one or more sections based on the WCS
along that axis can be selected, but for a regular selection
(e.g. 30:50:2) it will do the right thing.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fP
Input filename, in \fIimage(5NEMO)\fP format.
.br
[no default]
.TP
\fBout=\fP
Output filename
.br
[no default]
.TP
\fBzvar=x|y|z\fP
Slice variable (x, y, or z), along which axis the cube is sliced. This
axis will then become the 3rd axis.
.br
[\fBz\fP].
.TP
\fBzrange=\fP
Slices to select. Multiple planes can be selected, but must be in the
range 1..N\fBzvaR\fP. If none given, all planes will be selected.
Any \fInemoinp(1NEMO)\fP expression can be given, but only regular
ranges will result in a correct coordinate system.
.br
If a (single) negative value is used, the axis is decimated by the absolute
value of this number.
.br
Default: the whole \fIzvar\fP axis is extracted.
.TP
\fBzslabs=\fP\fIz1,z2,...\fP
Pairs of a Zmin,Zmax pairs in WCS to be selected.
.br
Default: not used.
.TP
\fBzscale=\fP
Scaling applied to \fBzslabs\fP to match the WCS in the header.
.br
Default: 1
.TP
\fBselect=t|f\fP
A boolean to signify of the selected planes are selected for output (t)
or the ones to ignore for output (f) - Not Implemented yet.

.SH "EXAMPLES"
Here is a cube, in which only the 2nd plane (z=1) is selected, the output
'cube' is hence two-dimensional
.nf

    ccdslice cube.in cube.out z 2

.fi
here is a cube which is re-oriented in YZX:
.nf

    ccdslice cube.in cube.out x

.fi
and here a somewhat cumbersome three-step process to select a 
subset of the X axis, but leaving the cube intact:
.nf

    ccdslice cube.in - x 2:10:2 | ccdslice - - x | ccdslice - cube.out x

.fi
and the same for selecting a subset of the Y axis, now in two steps:
.fi

    ccdslice cube.in - y 2:10:2 | ccdslice - cube.out y

.fi
Here an example to select some ranges in Z, where the input FITS file is in m/s,
but the selection is done in km/s. After this a noise map is computed for the selected
planes:
.nf

    fitsccd cube.fits - | ccdslice - - zslabs=100,200,500,600 zscale=1000 | ccdmom - cube.mom2 mom=-2
    
.fi
Here is an example of smoothing a cube with a boxcar of 6 channels, and decimating the cube by 6:
.nf

    fitsccd cube.fits - | ccdsmooth - - dir=z smooth=1/6::6 | ccdslice - cube6s.ccd zrange=-6

.fi

.SH "CAVEATS"
This code depends on the fact that the data is 3-dimensional, i.e. an
image cube. This is a limitation of the \fIimage(5NEMO)\fP format.
.PP
The coordinate system description of the new \fIzvar\fP axis can easily 
be corrupted by choosing an irregular layout with \fBxrange=\fP.

.SH "SEE ALSO"
ccdsub(1NEMO), ccdfits(1NEMO),  image(5NEMO), fits(5NEMO)

.SH "AUTHOR"
Peter Teuben

.SH "UPDATE HISTORY"
.nf
.ta +1.5i +5.5i
6-May-95	V1.0 Created		PJT
27-feb-2021	V1.2  add zslabs= and zscale=	PJT
11-may-2023	V1.3  better WCS handling	PJT
26-aug-2024	V1.4  allow decimation with negative zrange	PJT
.fi


src/image/io/ccdslice.c
/* 
 *	CCDSLICE: take slices from a cube
 *
 *	 6-may-95   V1.0 created 	Peter Teuben
 *      29-dec-01   V1.0a   also compute MapMin/Max
 *      27-feb-2021 V1.2    zslabs= and zscale implemented
 *       1-may-2022 V1.3    fix WCS for sampling
 */


#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>

#include <image.h>

string defv[] = {	/* keywords + help string for user interface */
    "in=???\n       Input filename (image)",
    "out=???\n      Output filename (image)",
    "zvar=z\n       Slice variable (x,y,z)",
    "zrange=\n      Slices to select (1..n) (<0 for every zrange'th)",
    "zslabs=\n      Zmin,Zmax pairs in WCS to select",
    "zscale=1\n     Scaling applied to zslabs",
    "select=t\n     Select the planes for output (t) or de-select those (f)",
    "VERSION=1.4\n  26-aug-2023 PJT",
    NULL,
};

string usage="takes slices from a cube";


#define X_SLICE 0
#define Y_SLICE 1
#define Z_SLICE 2

void slice(imageptr i, imageptr o, int mode, int *planes);
int  get_planes(int nslabs, real *slabs, int *planes, int na, real aref, real amin, real da, real ascale);

void nemo_main(void)
{
    imageptr iptr=NULL, optr=NULL;
    stream instr, outstr;
    int *planes, mode=-1, i, j;
    real *slabs, zscale;
    string zvar;
    int nx=0, ny=0, nz=0, maxplane=0;
    bool Qsel = getbparam("select");

    instr = stropen (getparam("in"),"r");
    read_image (instr,&iptr);            
    strclose(instr);                     

    if (Axis(iptr)) warning("axis=1 not fully supported yet");
    if (!Qsel) error("select=f not implemented yet");

    zvar = getparam("zvar");
    if (streq(zvar,"x")) {
        mode = X_SLICE;
        nx = Ny(iptr);
        ny = Nz(iptr);
        nz = Nx(iptr);
	maxplane = nx;
    } else if (streq(zvar,"y")) {
        mode = Y_SLICE;
        nx = Nx(iptr);
        ny = Nz(iptr);
        nz = Ny(iptr);
	maxplane = ny;
    } else if (streq(zvar,"z")) {
        mode = Z_SLICE;
        nx = Nx(iptr);
        ny = Ny(iptr);
        nz = Nz(iptr);
	maxplane = nz;	
    } else 
        error("Illegal slice axis %s, must be x,y,z",zvar);
    dprintf(1,"slice mode %d", mode);

    planes = (int *)  allocate(maxplane * sizeof(int));
    slabs  = (real *) allocate(maxplane * sizeof(real));
    
    zscale = getrparam("zscale");

    if (hasvalue("zrange")) {
        // @todo   nz re-used  !!!
        nz = nemoinpi(getparam("zrange"),planes,maxplane);
        if (nz<1) 
            error("Bad syntax %d for zrange=%s",nz,getparam("zrange"));
	if (nz>1 && ABS(planes[1]-planes[0]) > 1)
	  warning("Stepping assumed to be regular as %d for WCS correction", planes[1]-planes[0]);
	if (nz==1 && planes[0] < 0) {
	  int dz = -planes[0];
	  warning("New option zrange=%d maxplane=%d", planes[0], maxplane);
	  for (i=0, j=0;  ;  i++, j+=dz) {
	    if (i >= maxplane || j >= maxplane) {
	      maxplane = i;
	      break;
	    }
	    //dprintf(0,"planes %d %d\n", i+1, j+1);
	    planes[i] = j+1;
	  }
	  dprintf(0,"New maxplane=%d :  %d to %d\n", maxplane, planes[0], planes[maxplane-1]);
	  nz = maxplane;
	}
    } else if (hasvalue("zslabs")) {
        nz = nemoinpr(getparam("zslabs"),slabs,maxplane);
        if (nz<1) 
            error("Bad syntax %d for zrange=%s",nz,getparam("zslabs"));
	if (mode == X_SLICE)
	  nz = get_planes(nz, slabs, planes, Nx(iptr), Xref(iptr), Xmin(iptr), Dx(iptr), zscale);
	else if (mode == Y_SLICE)
	  nz = get_planes(nz, slabs, planes, Ny(iptr), Yref(iptr), Ymin(iptr), Dy(iptr), zscale);
	else if (mode == Z_SLICE) 
	  nz = get_planes(nz, slabs, planes, Nz(iptr), Zref(iptr), Zmin(iptr), Dz(iptr), zscale);
	else
	  error("mode not implemented for slabs");
    } else {
        for (i=0; i<nz; i++)
            planes[i] = i+1;
    }
    create_cube(&optr,nx,ny,nz);
    copy_header(iptr,optr,1);
    slice(iptr,optr,mode,planes);

    outstr = stropen(getparam("out"),"w");
    write_image(outstr,optr);
    strclose(outstr);
}


/* from miriad::imbin WCS correction:
 *  crpixo(i) = 1 + (crpixi(i)-blc(i))/bin(1,i)
 *  cdelto(i) = bin(1,i) * cdelti(i)
 *  crpixo(i) = crpixo(i) - 0.5*(bin(2,i)-1)/bin(2,i)
 *
 *  where bin(1) = step  and bin(2) = width
 *
 */


void slice(imageptr i, imageptr o, int mode, int *planes)
{
    int x, y, z, iz;
    real ival;


    if (mode==X_SLICE) {
        warning("Code for X not converted to fix reference pixel value");
        for(iz=0; iz<Nz(o); iz++) {
            z = planes[iz]-1;
            if (z<0 || z>Nx(i)) 
                error("%d: illegal plane in x, max is %d",z,Nx(i));
            for (y=0; y<Ny(o); y++)
	      for (x=0; x<Nx(o); x++) {
		ival = CubeValue(i,z,x,y);
                CubeValue(o,x,y,iz) = ival;
	      }
        }
        Namex(o) = Namey(i);
        Namey(o) = Namez(i);
        Namez(o) = Namex(i);
        Xmin(o) = Ymin(i);  // fix
        Ymin(o) = Zmin(i);
        Zmin(o) = Xmin(i);
        Dx(o) = Dy(i);      // fix
        Dy(o) = Dz(i);
        Dz(o) = Dx(i);
    } else if (mode==Y_SLICE) {
        warning("Code for Y not converted to fix reference pixel value");      
        for(iz=0; iz<Nz(o); iz++) {
            z = planes[iz]-1;
            if (z<0 || z>Ny(i)) 
                error("%d: illegal plane in y, max is %d",z,Ny(i));
            for (y=0; y<Ny(o); y++)
	      for (x=0; x<Nx(o); x++) {
		ival = CubeValue(i,x,z,y);
                CubeValue(o,x,y,iz) = ival;
	      }
        }
        Namex(o) = Namex(i);
        Namey(o) = Namez(i);
        Namez(o) = Namey(i);
        Xmin(o) = Xmin(i);
        Ymin(o) = Zmin(i);
        Zmin(o) = Ymin(i);
        Dx(o) = Dx(i);
        Dy(o) = Dz(i);
        Dz(o) = Dy(i);
    } else if (mode==Z_SLICE) {
        for(iz=0; iz<Nz(o); iz++) {
            z = planes[iz]-1;
            if (z<0 || z>Nz(i)) 
                error("%d: illegal plane in z, max is %d",z,Nz(i));
            for (y=0; y<Ny(o); y++)
	      for (x=0; x<Nx(o); x++)
                CubeValue(o,x,y,iz) = CubeValue(i,x,y,z);
        }
	real width_step = planes[1]-planes[0];
	dprintf(0,"Z_SLICE: %d %g\n",planes[0],width_step);
	Zref(o) = (Zref(i)-planes[0])/width_step;
	Dz(o) = Dz(i) * width_step;
	Zref(o) = Zref(o) - 0.5*(width_step - 1.0)/width_step;
    }
    minmax_image(o);
}

/*
 *  convert slabs to 1..N based index array
 *
 */

int get_planes(int nslabs, real *slabs, int *planes, int na, real aref, real amin, real da, real ascale)
{
  int i0, i1, i, j, nz = 0;
  real axis_min, axis_max;

  dprintf(0,"Axis: %d %g  %g %g\n",na,aref,amin,da);
  axis_min =       -aref*da + amin;
  axis_max = (na-1-aref)*da + amin;
  dprintf(0,"Axis min/max: %g %g\n",axis_min,axis_max);
  
  if (nslabs%2) error("Not an even number of slab values");

  for (i=0; i<nslabs; i+=2) {
    i0 = (int) ((ascale*slabs[i]  -amin)/da + aref);
    i1 = (int) ((ascale*slabs[i+1]-amin)/da + aref);
    if (i0<0) {
      warning("%d < 0: ",i0);
      i0=0;
    }
    if (i1>=na) {
      warning("%d >= %d: ",i1,na);
      i1=na-1;
    }
    dprintf(0,"slab %g %g -> %d %d\n",slabs[i],slabs[i+1],i0,i1);
    if (i0<i1)
      for (j=i0; j<=i1; j++)
	planes[nz++] = j+1;    
    else
      for (j=i1; j<=i0; j++)
	planes[nz++] = j+1;
  }
  dprintf(0,"Total planes: %d\n",nz);
  return nz;
}
  
